### PR TITLE
vNext Third Party Plugin Architecture - WIP/POC

### PR DESCRIPTION
**Pull request by JamesAPetts**
EDIT:
This blew up a little... take a deep breath.

Docs are more notes at this point, and could use some tidying up. New vNext plugin architecture, highly experimental.

[Live demo of HelloWorldMouseTool plugin.](https://deploy-preview-605--cornerstone-tools.netlify.com/examples/plugin/)

The plugin contains one mixin, module and tool. The tool logs some data from the module and its mixin to the console on click. The source is available at `examples/plugin`.

Tools, mixins and modules can now be registered to a `cornerstoneTools` instance via these commands:

```
const cTools = cornerstoneTools.init();

cTools.thirdParty.registerTool(ExampleTool, 'Example');
cTools.thirdParty.registerModule(exampleModule, 'exampleModule');
cTools.thirdParty.registerMixin(exampleMixin, 'exampleMixin);
```
Note: Mixins are very limited in scope for now, as will become evident from the docs.

Also if you have a bunch of tools, mixins and modules that are related/depend on each other, you can stick them all in a plugin with the following schema:

```
const samplePlugin = {
  //Mandatory
  name: 'samplePlugin',

  // Optional
  mixins: [
    {
      mixin, // Mixin Object
      name   // Name of the mixin
    },
    ...
  ],

  // Optional
  modules: [
    {
      module, // Module Object
      name  // Name of the module
    },
    ...
  ],

  // Optional
  tools: [
    {
      tool:  // The tool Class / Constructor.
      name: // Name of the Tool.
    }
  ]

};
```

This plugin can then be registered like so:

```
const cTools = cornerstoneTools.init();

cTools.thirdParty.registerPlugin(samplePlugin);
```

Mixins and modules are plugged straight into `cornerstoneTools` ready for use. Tools from plugins can be enabled once registered by:

```
cTools.addTool(cTools.thirdParty.tools.HelloWorldMouseTool);
```

Please read the 'Third Party Addons' docs included for more info.
Feel free to give any feedback, this is very much a first draft/POC.

### CODE REVIEW

File: docs/latest/custom-tools/choosing-a-base-class.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+## Choosing A Base Class {#choosing-a-base-class}
+
+There are 3 base classes to choose from when building a tool:
```
Comment (dannyrb) [inline docs/latest/custom-tools/choosing-a-base-class.md]: While we see it as a "base class", implementers might understand this better if we refer to them as different tool types. Each comes with different hooks/callbacks, and afford different default functionality to make creating different types of tools easier. A table of the callbacks the tool type offers, how to extend/use/implement it, and when/why you would use that tool type are all useful bits of information

File: docs/latest/custom-tools/creating-your-tool.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+By convention the class name should be in PascalCase, and suffixed with:
+- `Tool` - If it supports both `mouse` and `touch` input.
+- `MouseTool` - If it only supports `mouse` input.
+- `TouchTool` - If it only supports `touch` input.
+
+For example, our example is going to support mouse, so we shall call it the `HelloWorldMouseTool`:
+
+```js
+import external from './../externalModules.js';
+import BaseTool from './../base/BaseTool.js';
```
Comment (dannyrb) [inline docs/latest/custom-tools/creating-your-tool.md]: The imports only work internally. We need to identify the best way to access, extend/implement these interfaces externally.

Comment (dannyrb) [reply docs/latest/custom-tools/creating-your-tool.md]: Quilljs has an interesting approach:

_Example of extending a "base class"_
```js
// https://github.com/quilljs/parchment/#blots
import Quill from 'quill'
let Inline = Quill.import('blots/inline')

class SpellingBlot extends Inline {
  static create(value) {
    let node = super.create(value)
    node.setAttribute('data-invalid-spelling', '')
    node.setAttribute('data-alternative-spellings', [])
    return node
  }

  static formats(domNode) {
    return {
      isInvalidSpelling: domNode.hasAttribute('data-invalid-spelling'),
      alternativeSpellings: domNode.getAttribute('data-alternative-spellings'),
    }
  }
}

SpellingBlot.blotName = 'spelling'
SpellingBlot.className = 'spelling'
SpellingBlot.tagName = 'SPAN'

export default SpellingBlot
```

And then to add it... [register](https://quilljs.com/docs/api/#register)
```js
Quill.register(SpellingBlot, true);
```

File: docs/latest/custom-tools/event-dispatcher-callbacks.md [hunk 3]
```diff
@@ -1,3 +1,35 @@
 ## Event Dispatcher Callbacks {#event-dispatcher-callbacks}
 
-_This section needs content_
+Here we can add the meat of our tool. Event dispatchers check for methods on tools and fire them when appropriate [TODO: List them all?? Or just link to api? Not sure yet. ]
```
Comment (dannyrb) [inline docs/latest/custom-tools/event-dispatcher-callbacks.md]: An image that describing event order, how tool's are chosen to be dispatched to, the methods that are associated with each event, and  `isToolLocked` would be super handy here. I'll look in to generating something for this?

File: docs/latest/custom-tools/adding-mixins.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import external from './../externalModules.js';
+import BaseTool from './../base/BaseTool.js';
+
+export default class HelloWorldMouseTool extends BaseTool {
+  constructor (name = 'HelloWorldMouse') {
+    super({
+      name,
+      supportedInteractionTypes: ['mouse'],
+      mixins = [
+        'activeOrDisabledBinaryTool'
```
[the remark this answers is not in the document]
Comment (JamesAPetts) [reply docs/latest/custom-tools/adding-mixins.md]: Yeah, the plugin framework (or lack there of) does need some work, I agree its not blocking though, the framework can develop over the course of v3xx.

File: docs/latest/custom-tools/event-dispatcher-callbacks.md (hunk 3 above)
Comment (JamesAPetts) [reply docs/latest/custom-tools/event-dispatcher-callbacks.md]: That would honestly be super neat.

File: docs/latest/custom-tools/creating-your-tool.md (hunk 2 above)
Comment (JamesAPetts) [reply docs/latest/custom-tools/creating-your-tool.md]: Yeah, we need to expose a load of this stuff. So far the `BaseTool`s aren't even exposed at the surface level. I'm using the doc writing as a way of identifying what'd be needed.

File: docs/latest/custom-tools/choosing-a-base-class.md (hunk 1 above)
Comment (JamesAPetts) [reply docs/latest/custom-tools/choosing-a-base-class.md]: Yeah the semantics could use a little work, maybe we should call these 'Base Tools' colloquially?

File: docs/latest/custom-tools/event-dispatcher-callbacks.md (hunk 3 above)
Comment (dannyrb) [reply docs/latest/custom-tools/event-dispatcher-callbacks.md]: Created a note here: https://github.com/cornerstonejs/cornerstoneTools/projects/2#card-13374470

Feel free to do the same for anything you want to pass off to someone else or make a note of. Comments can be easily forgotten after a merge.

Comment (JamesAPetts) [reply docs/latest/custom-tools/event-dispatcher-callbacks.md]: Sure, I more wrote that comment to myself haha, planned to continue this morning. But yeah I'll post any discussion/issues back to the board.

File: docs/latest/custom-tools/adding-mixins.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (JamesAPetts) [reply docs/latest/custom-tools/adding-mixins.md]: Or maybe it can now ¯\\_(ツ)\_/¯.

**Comment by JamesAPetts** [comment]
Hey @pieper , I didn't get that far on Friday. I'm going to make a `hello world` plugin with one tool, mixin and module, and stick it in an example on Monday 👍. I'll post in slack when it's up.

**Comment by JamesAPetts** [comment]
@pieper I made a simple plugin that logs to the console on click. The plugin contains one mixin, module and tool, that all interact.

**Comment by JamesAPetts** [comment]
We are going to merge this for now, however this framework will surely develop with time. We will review the plugin framework at RC-1 to see if any major revisions are needed before release.

**Pull request merged**